payu/branch.py

```diff
@@ -0,0 +1,203 @@
+        print(f"Removed {sym_dir} symlink to {previous_path}")
+
+    # Create symlink, if experiment directory exists in laboratory
+    if dir_path.exists():
+        sym_path.symlink_to(dir_path)
+        print(f"Added {sym_dir} symlink to {dir_path}")
+
+
+def clone(repository: str,
+          directory: Path,
+          branch: Optional[Path] = None,
+          new_branch_name: Optional[str] = None,
+          keep_uuid: bool = False,
+          model_type: Optional[str] = None,
+          config_path: Optional[Path] = None,
+          lab_path: Optional[Path] = None,
+          restart_path: Optional[Path] = None) -> None:
+    """Clone an experiment control repo"""
+    # git clone the repository
+    git_clone(repository, directory, branch)
+
+    # Resolve directory to an absolute path and cd into cloned directory
+    previous_directory = os.getcwd()
+    dir_path = directory.resolve()
+    os.chdir(directory)
+
+    # Initial lab and metadata
+    lab = Laboratory(model_type, config_path, lab_path)
+
+    # Use checkout wrapper
+    if new_branch_name is not None:
+        # Create and checkout new branch
+        checkout_branch(lab=lab,
+                        is_new_branch=True,
+                        branch_name=new_branch_name,
+                        restart_path=restart_path,
+                        config_path=config_path)
+    else:
+        # Checkout branch
+        if branch is None:
+            branch = get_git_branch(dir_path)
+
+        checkout_branch(lab=lab,
+                        branch_name=branch,
+                        config_path=config_path,
+                        is_new_experiment=not keep_uuid,
+                        restart_path=restart_path)
+        # Note: is_new_experiment ensures new uuid and metadata is created
+        # Otherwise uuid is generated only if there's no pre-existing uuid
+
+    # Change back to previous directory
+    os.chdir(previous_directory)
+
+
+def print_branch_metadata(branch: git.Head, verbose: bool = False):
+    """Print uuid for each branch. If verbose is true, it will print all lines
+    of the metadata file"""
+    contains_metadata = False
+    # Note: Blobs are files in the commit tree
+    for blob in branch.commit.tree.blobs:
+        if blob.name == 'metadata.yaml':
+            contains_metadata = True
+
+            # Read file contents
+            content = blob.data_stream.read().decode('utf-8')
+            if verbose:
+                # Print all metadata
+                for line in content.splitlines():
+                    print(f'    {line}')
+            else:
+                # Print uuid
+                metadata = YAML().load(content)
+                uuid = metadata.get('uuid', None)
+                if uuid is not None:
+                    print(f"    uuid: {uuid}")
+                else:
+                    print(f"    No uuid in metadata file")
+
+    if not contains_metadata:
+        print("    No metadata file found")
+
```

Comment:
At this stage, I've just added a "No config file found" tag, so the branch name is still there. I'll add returning metadata to the function as a future thing to do. I don't know much caching in python programs either so that'll be something to learn!